translator/internaldata/oc_to_metrics.go

```diff
@@ -0,0 +1,315 @@
+	// which will gets its Resource field from "md.Resource".
+	//
+	// We will end up with with one or more ResourceMetrics like this:
+	//
+	// ResourceMetrics           ResourceMetrics  ResourceMetrics
+	// +-------+-------+---+-------+ +--------------+ +--------------+
+	// |Metric1|Metric2|...|MetricM| |Metric        | |Metric        | ...
+	// +-------+-------+---+-------+ +--------------+ +--------------+
+
+	// Count the number of metrics that have nil Resource and need to be combined
+	// in one slice.
+	combinedMetricCount := 0
+	distinctResourceCount := 0
+	for _, ocMetric := range md.Metrics {
+		if ocMetric == nil {
+			// Skip nil metrics.
+			continue
+		}
+		if ocMetric.Resource == nil {
+			combinedMetricCount++
+		} else {
+			distinctResourceCount++
+		}
+	}
+	// Total number of resources is equal to:
+	// 1 (for all metrics with nil resource) + numMetricsWithResource (distinctResourceCount).
+	rms := metricData.ResourceMetrics()
+	rms.Resize(distinctResourceCount + 1)
+	ocNodeResourceToInternal(md.Node, md.Resource, rms.At(0).Resource())
+
+	// Allocate a slice for metrics that need to be combined into first ResourceMetrics.
+	ilms := rms.At(0).InstrumentationLibraryMetrics()
+	ilms.Resize(1)
+	combinedMetrics := ilms.At(0).Metrics()
+	combinedMetrics.Resize(combinedMetricCount)
+
+	// Now do the metric translation and place them in appropriate ResourceMetrics
+	// instances.
+
+	// Index to next available slot in "combinedMetrics" slice.
+	combinedMetricIdx := 0
+	// First resourcemetric is used for the default resource, so start with 1.
+	resourceMetricIdx := 1
+	for _, ocMetric := range md.Metrics {
+		if ocMetric == nil {
+			// Skip nil metrics.
+			continue
+		}
+
+		if ocMetric.Resource == nil {
+			// Add the metric to the "combinedMetrics". combinedMetrics length is equal
+			// to combinedMetricCount. The loop above that calculates combinedMetricCount
+			// has exact same conditions as we have here in this loop.
+			ocMetricToInternal(ocMetric, combinedMetrics.At(combinedMetricIdx))
+			combinedMetricIdx++
+		} else {
+			// This metric has a different Resource and must be placed in a different
+			// ResourceMetrics instance. Create a separate ResourceMetrics item just for this metric.
+			ocMetricToResourceMetrics(ocMetric, md.Node, metricData.ResourceMetrics().At(resourceMetricIdx))
+			resourceMetricIdx++
+		}
+	}
+
+	return metricData
+}
+
+func ocMetricToResourceMetrics(ocMetric *ocmetrics.Metric, node *occommon.Node, out data.ResourceMetrics) {
+	ocNodeResourceToInternal(node, ocMetric.Resource, out.Resource())
+	ilms := out.InstrumentationLibraryMetrics()
+	ilms.Resize(1)
+	metrics := ilms.At(0).Metrics()
+	metrics.Resize(1)
+	ocMetricToInternal(ocMetric, metrics.At(0))
+}
+
+// ocMetricToInternal conversts ocMetric to internal representation and fill metric
+func ocMetricToInternal(ocMetric *ocmetrics.Metric, metric data.Metric) {
+	metric.MetricDescriptor().InitEmpty()
```

Comment:
Please use temp var for `metric.MetricDescriptor()` to execute it once only. It is a low cost function but not zero cost. Similarly elsewhere where we call accessor functions.

Reply:
done